lib/functions/OverloadTree.ts

```diff
@@ -54,23 +77,49 @@ export class OverloadTree {
   }
 
   private _addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
-    const argumentType = argumentTypes.shift();
+    const [ argumentType, ..._argumentTypes ] = argumentTypes;
     if (!argumentType) {
       this.implementation = func;
       return;
     }
     if (!this.subTrees[argumentType]) {
-      this.subTrees[argumentType] = new OverloadTree();
+      this.subTrees[argumentType] = new OverloadTree(this.depth + 1);
+    }
+    this.subTrees[argumentType]._addOverload(_argumentTypes, func);
+    // Defined by https://www.w3.org/TR/xpath-31/#promotion .
+    //  When a function takes a string, it can also accept a XSD_ANY_URI if it is cased first.
+    if (argumentType === TypeURL.XSD_STRING) {
+      this.addPromotedOverload(TypeURL.XSD_ANY_URI, func, arg => string(arg.str()), _argumentTypes);
     }
-    this.subTrees[argumentType]._addOverload(argumentTypes, func);
+    if (argumentType === TypeURL.XSD_FLOAT) {
+      this.addPromotedOverload(TypeURL.XSD_DOUBLE, func, arg =>
+        number((<E.NumericLiteral>arg).typedValue, TypeURL.XSD_DOUBLE), _argumentTypes);
+    }
+    if (argumentType === TypeURL.XSD_DECIMAL) {
+      this.addPromotedOverload(TypeURL.XSD_FLOAT, func, arg =>
+        number((<E.NumericLiteral>arg).typedValue, TypeURL.XSD_FLOAT), _argumentTypes);
+      this.addPromotedOverload(TypeURL.XSD_DOUBLE, func, arg =>
+        number((<E.NumericLiteral>arg).typedValue, TypeURL.XSD_DOUBLE), _argumentTypes);
+    }
+  }
+
+  private addPromotedOverload(typeToPromote: ArgumentType, func: E.SimpleApplication,
+    conversionFunction: (arg: E.TermExpression) => E.TermExpression, argumentTypes: ArgumentType[]): void {
+    if (!this.subTrees[typeToPromote]) {
+      this.subTrees[typeToPromote] = new OverloadTree(this.depth + 1);
+    }
+    this.subTrees[typeToPromote]._addOverload(argumentTypes, args => func([
+      ...args.slice(0, this.depth),
+      conversionFunction(args[this.depth]),
+      ...args.slice(this.depth + 1, args.length),
+    ]));
   }
```

Comment:
This code should allow type promotion as talked about in https://www.w3.org/TR/xpath-31/#promotion . At least, how I understand it. If this is wrong please let me know.

Reply:
At first glance it looks like it could work.
But we'll now for certain once we have tests covering these cases.

Reply:
Yes, I will add the scheme first and then start writing the tests. With 'wrong' I meant the interpretation of 'promotion'. :)

Reply:
A special thing to consider here is that subtype substitution and type promotion can work at the same time. There certainly should be tests for that, as I'm not sure the code handles this like it is (but it's also likely that I'm just missing something).

Reply:
I pretty sure this wouldn't be a problem since promotion is always from a non super type. To another type. And of course like it is now, having a restriction of anyURI provided to a function expecting an xsd_string, the restriction will be substituted to anyUri and will then be promoted to xsd_string.

Reply:
I'm still convinced we can't really test this. So unless someone can give an example on when this could create a problem I consider this to be done.

Reply:
I added some documentation so people don't break the system and also a test that first performs substitution and that promotion.